### PR TITLE
Breaking: runtime-deprecation on '~/.eslintrc' (refs eslint/rfcs#32)

### PR DESCRIPTION
**Pull request by mysticatea**
> Don't merge until we finished releasing the last one of 6.x.

**What is the purpose of this pull request? (put an "X" next to item)**

[X] Add something to the core

**What changes did you make? (Give an overview)**

This PR adds runtime-deprecation about personal config files (`~/.eslintrc.*`). This is a part of RFC32.

ESLint emits deprecation warnings when:

- when ESLint loaded a personal config file.
- when ESLint ignored a personal config file because of project config files that *don't* have `root:true`.

The warning messages are like:

```
(node:20312) [ESLINT_PERSONAL_CONFOG_LOAD] DeprecationWarning: The '~/.eslintrc.*' config file has been deprecated. Please use config files for each project or '--config' option. (found in "eslint\fixtures\config-hierarchy\personal-config\home-folder\.eslintrc.json")
```

```
(node:20312) [ESLINT_PERSONAL_CONFOG_SUPPRESS] DeprecationWarning: The '~/.eslintrc.*' config file has been deprecated. Please remove it or add 'root:true' into the config file of your projects in order to avoid loading '~/.eslintrc.*' accidentally. (found in "eslint\fixtures\config-hierarchy\personal-config\home-folder\.eslintrc.json")
```

**Is there anything you'd like reviewers to focus on?**

The warning messages are understandable?

### CODE REVIEW

File: lib/cli-engine/cascading-config-array-factory.js [hunk 1]
```diff
@@ -335,6 +329,22 @@ class CascadingConfigArrayFactory {
...
+
+        // Consider this is root.
+        if (directoryPath === homePath && cwd !== homePath) {
+            debug("Stop traversing because of considered root.");
+            if (configsExistInSubdirs && configArray.length > 0) {
+                const lastElement = configArray[configArray.length - 1];
+
+                emitDeprecationWarning(
+                    lastElement.filePath,
+                    "ESLINT_PERSONAL_CONFOG_SUPPRESS"
```
Comment (kaicataldo) [inline lib/cli-engine/cascading-config-array-factory.js]:
```suggestion
                    "ESLINT_PERSONAL_CONFIG_SUPPRESS"
```

File: lib/cli-engine/cascading-config-array-factory.js [hunk 2]
```diff
@@ -402,10 +415,22 @@ class CascadingConfigArrayFactory {
...
+                    { name: "PersonalConfig" }
                 );
+
+                if (personalConfigArray.length > 0) {
+                    const lastElement =
+                        personalConfigArray[personalConfigArray.length - 1];
+
+                    emitDeprecationWarning(
+                        lastElement.filePath,
+                        "ESLINT_PERSONAL_CONFOG_LOAD"
```
Comment (kaicataldo) [inline lib/cli-engine/cascading-config-array-factory.js]:
```suggestion
                        "ESLINT_PERSONAL_CONFIG_LOAD"
```

File: lib/shared/deprecation-warnings.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+//------------------------------------------------------------------------------
+// Private
+//------------------------------------------------------------------------------
+
+// Defitions for deprecation warnings.
+const deprecationWarningMessages = {
+    ESLINT_LEGACY_ECMAFEATURES:
+        "The 'ecmaFeatures' config file property is deprecated, and has no effect.",
+    ESLINT_PERSONAL_CONFOG_LOAD:
```
Comment (kaicataldo) [inline lib/shared/deprecation-warnings.js]:
```suggestion
    ESLINT_PERSONAL_CONFIG_LOAD:
```

File: lib/shared/deprecation-warnings.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        "The '~/.eslintrc.*' config file has been deprecated. " +
+        "Please use config files for each project or '--config' option.",
+    ESLINT_PERSONAL_CONFOG_SUPPRESS:
```
Comment (kaicataldo) [inline lib/shared/deprecation-warnings.js]:
```suggestion
    ESLINT_PERSONAL_CONFIG_SUPPRESS:
```

File: lib/shared/deprecation-warnings.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+const lodash = require("lodash");
+
+//------------------------------------------------------------------------------
+// Private
+//------------------------------------------------------------------------------
+
+// Defitions for deprecation warnings.
+const deprecationWarningMessages = {
+    ESLINT_LEGACY_ECMAFEATURES:
+        "The 'ecmaFeatures' config file property is deprecated, and has no effect.",
```
Comment (kaicataldo) [inline lib/shared/deprecation-warnings.js]:
```suggestion
        "The 'ecmaFeatures' config file property is deprecated and has no effect.",
```

File: lib/shared/deprecation-warnings.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    ESLINT_PERSONAL_CONFOG_LOAD:
+        "The '~/.eslintrc.*' config file has been deprecated. " +
+        "Please use config files for each project or '--config' option.",
```
Comment (kaicataldo) [inline lib/shared/deprecation-warnings.js]:
```suggestion
        "Please use a config file per project or the '--config' option.",
```

File: lib/shared/deprecation-warnings.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+//------------------------------------------------------------------------------
+// Private
+//------------------------------------------------------------------------------
+
+// Defitions for deprecation warnings.
+const deprecationWarningMessages = {
+    ESLINT_LEGACY_ECMAFEATURES:
+        "The 'ecmaFeatures' config file property is deprecated, and has no effect.",
+    ESLINT_PERSONAL_CONFOG_LOAD:
+        "The '~/.eslintrc.*' config file has been deprecated. " +
```
Comment (kaicataldo) [inline lib/shared/deprecation-warnings.js]:
```suggestion
        "'~/.eslintrc.*' config files have been deprecated. " +
```

File: lib/shared/deprecation-warnings.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        "Please use config files for each project or '--config' option.",
+    ESLINT_PERSONAL_CONFOG_SUPPRESS:
+        "The '~/.eslintrc.*' config file has been deprecated. " +
```
Comment (kaicataldo) [inline lib/shared/deprecation-warnings.js]:
```suggestion
        "'~/.eslintrc.*' config files have been deprecated. " +
```

File: lib/shared/deprecation-warnings.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        "Please remove it or add 'root:true' into the config file of your " +
```
Comment (kaicataldo) [inline lib/shared/deprecation-warnings.js]:
```suggestion
        "Please remove it or add 'root:true' to the config files in your " +
```

**Comment by mysticatea** [comment]
Thank you for your review. I have updated this PR.

**Comment by mysticatea** [comment]
I added some tests. It clarifies that ESLint raises new deprecation warnings only when the cases that ESLint change the behavior after the personal config file feature were removed.

**Pull request merged**